contribute/architecture/state.md

```diff
@@ -0,0 +1,142 @@
+
+[Services](services.md) don't use XORM directly. Instead, services use the _SQL store_, a special type of service that provides an abstraction for the database layer. There are two ways of using the `sqlstore`: using `sqlstore` handlers, and using the `SqlStore` instance.
+
+## `sqlstore` handlers
+
+> **Deprecated:** We are deprecating `sqlstore` handlers in favor of using the `SqlStore`  object directly in each service. Since most services still use the `sqlstore` handlers, we still want to explain how they work.
+
+The `sqlstore` package allows you to register command handlers that either store, or retrieve objects from the database. `sqlstore` handlers are similar to services:
+
+- Services are command handlers that _contain business logic_.
+- `sqlstore` handlers are command handlers that _access the database_.
+
+More importantly, command handlers for services are _methods_, while `sqlstore` handlers are _functions_ that access global variables.
+
+### Register a `sqlstore` handler
+
+> **Deprecated:** Refer to the [deprecation note for `sqlstore` handlers](#sqlstore-handlers).
+
+To register a handler:
+
+- Create a new file `myrepo.go` in the `sqlstore` package.
+- Create a [command handler](services.md#handle-commands).
+- Register the handler in the `init` function:
+
+```go
+func init() {
+    bus.AddHandler("sql", SaveDashboard)
+}
+
+func DeleteDashboard(cmd *models.DeleteDashboardCommand) error {
+    return inTransaction(func(sess *DBSession) error {
+        _, err := sess.Exec("DELETE FROM dashboards WHERE dashboard_id=?", cmd.DashboardID)
+        return err
+    })
+}
+```
+
+Here, `inTransaction` is a helper function in the `sqlstore` package that provides a session, that lets you execute SQL statements.
+
+## `SqlStore`
+
+As opposed to a `sqlstore` handler, the `SqlStore` is a service itself. The `SqlStore` has the same responsibility however: to store and retrieve objects, to and from the database.
+
+To use the `SqlStore`, inject the `SQLStore` in your service struct:
+
+```go
+type MyService struct {
+    SQLStore *sqlstore.SqlStore `inject:""`
+}
+```
+
+You can now make SQL queries in any of your [command handlers](services.md#handle-commands) or [event listeners](services.md#subscribe-to-events):
+
+```go
+func (s *MyService) DeleteDashboard(cmd *models.DeleteDashboardCommand) error {
+    if err := s.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
+        _, err := sess.Exec("DELETE FROM dashboards WHERE dashboard_id=?", cmd.DashboardID)
+        return err
+    })
+}
+```
+
+For transactions, use the `WithTransactionalDbSession` method instead.
+
+## Migrations
+
+As Grafana evolves, it becomes necessary to create _schema migrations_ for one or more database tables.
+
+To see all the types of migrations you can add, refer to [migrations.go](/pkg/services/sqlstore/migrator/migrations.go).
```

Comment:
```suggestion
Examples of migrations you can add, refer to [migrations.go](/pkg/services/sqlstore/migrator/migrations.go).
```

Reply:
Might be wrong as I did not double-check but I would be surprised if we cover every possible case.

Reply:
Are we encouraging people to write their own migration types (as opposed to migrations)?

These seem to be the available ones:

- AddColumnMigration
- AddIndexMigration
- DropIndexMigration
- AddTableMigration
- DropTableMigration
- RenameTableMigration
- CopyTableDataMigration
- TableCharsetMigration

I wanted to say that you can choose from any of these.

Reply:
Great, that makes sense. Also, seems like we `RawSqlMigration` which would effectively cover _every_ possible case. 


I'm fine with the current wording.